File: basin_connectivity_schematic.m
%% Schematic: basin connectivity at low vs high temperature
%  Left panel:  low T  — basins are narrow, target is isolated (retrieval)
%  Right panel: high T — basins widen, giant connected cluster (paramagnetic)

clear; close all;

rng(42);  % reproducible layout

%% Parameters
n_patterns = 14;                   % number of patterns
R = 1;                             % circle radius (represents S^{N-1})

% Place pattern centers at random angles on the circle
angles = sort(2*pi*rand(1, n_patterns));

% Basin half-widths (angular radius of spherical cap)
theta_low  = 0.28;    % low T:  narrow effective basins
theta_high = 0.50;    % high T: wider effective basins

target_idx = 1;       % the "target" pattern

%% Helper: check if two basins overlap
overlap = @(a1, a2, theta) ...
    min(abs(a1 - a2), 2*pi - abs(a1 - a2)) < 2*theta;

%% Helper: find connected component containing target via BFS
function comp = bfs_component(adj, start)
    n = size(adj, 1);
    comp = false(1, n);
    queue = start;
    comp(start) = true;
    while ~isempty(queue)
        node = queue(1); queue(1) = [];
        nbrs = find(adj(node, :) & ~comp);
        comp(nbrs) = true;
        queue = [queue, nbrs];
    end
end

%% Build adjacency matrices
adj_low  = false(n_patterns);
adj_high = false(n_patterns);
for i = 1:n_patterns
    for j = i+1:n_patterns
        if overlap(angles(i), angles(j), theta_low)
            adj_low(i,j) = true; adj_low(j,i) = true;
        end
        if overlap(angles(i), angles(j), theta_high)
            adj_high(i,j) = true; adj_high(j,i) = true;
        end
    end
end

% Connected components containing target
comp_low  = bfs_component(adj_low,  target_idx);
comp_high = bfs_component(adj_high, target_idx);

%% Colors
col_bg        = [0.92 0.92 0.92];  % background basin fill
col_target    = [0.85 0.25 0.25];  % target basin
col_connected = [0.30 0.65 0.90];  % connected to target
col_isolated  = [0.75 0.75 0.75];  % not connected
col_edge_on   = [0.20 0.50 0.80];  % active connection line
col_edge_off  = [0.85 0.85 0.85];  % inactive connection (shown faintly)
col_sphere    = [0.3 0.3 0.3];     % sphere outline

%% Draw one panel
function draw_panel(ax, angles, theta, adj, comp, target_idx, ...
                    col_target, col_connected, col_isolated, ...
                    col_edge_on, col_edge_off, col_sphere, title_str)
    axes(ax); hold on; axis equal off;
    n = length(angles);
    R = 1;

    % Draw sphere outline
    t = linspace(0, 2*pi, 300);
    plot(R*cos(t), R*sin(t), '-', 'Color', col_sphere, 'LineWidth', 2);

    % Draw basin arcs (as filled wedges on the circle)
    for i = 1:n
        a = angles(i);
        arc_t = linspace(a - theta, a + theta, 80);

        % Basin fill (arc + radial shading)
        r_inner = 0.82; r_outer = 1.18;
        x_fill = [r_inner*cos(arc_t), fliplr(r_outer*cos(arc_t))];
        y_fill = [r_inner*sin(arc_t), fliplr(r_outer*sin(arc_t))];

        if i == target_idx
            fc = col_target;
            fa = 0.45;
        elseif comp(i)
            fc = col_connected;
            fa = 0.35;
        else
            fc = col_isolated;
            fa = 0.25;
        end
        fill(x_fill, y_fill, fc, 'FaceAlpha', fa, 'EdgeColor', 'none');

        % Basin arc outline
        if i == target_idx
            plot(R*cos(arc_t), R*sin(arc_t), '-', 'Color', col_target, 'LineWidth', 3);
        elseif comp(i)
            plot(R*cos(arc_t), R*sin(arc_t), '-', 'Color', col_connected, 'LineWidth', 2);
        else
            plot(R*cos(arc_t), R*sin(arc_t), '-', 'Color', col_isolated, 'LineWidth', 1.5);
        end
    end

    % Draw connection edges (as chords)
    for i = 1:n
        for j = i+1:n
            if adj(i,j)
                xi = R*cos(angles(i)); yi = R*sin(angles(i));
                xj = R*cos(angles(j)); yj = R*sin(angles(j));
                if comp(i) && comp(j)
                    plot([xi xj], [yi yj], '-', 'Color', [col_edge_on 0.7], ...
                        'LineWidth', 2);
                else
                    plot([xi xj], [yi yj], '-', 'Color', [col_edge_off 0.5], ...
                        'LineWidth', 1);
                end
            end
        end
    end

    % Draw pattern centers as dots
    for i = 1:n
        xi = R*cos(angles(i)); yi = R*sin(angles(i));
        if i == target_idx
            plot(xi, yi, 'o', 'MarkerSize', 10, 'MarkerFaceColor', col_target, ...
                'MarkerEdgeColor', 'w', 'LineWidth', 1.5);
        elseif comp(i)
            plot(xi, yi, 'o', 'MarkerSize', 7, 'MarkerFaceColor', col_connected, ...
                'MarkerEdgeColor', 'w', 'LineWidth', 1);
        else
            plot(xi, yi, 'o', 'MarkerSize', 7, 'MarkerFaceColor', col_isolated, ...
                'MarkerEdgeColor', 'w', 'LineWidth', 1);
        end
    end

    % Label target
    xt = 1.30*cos(angles(target_idx));
    yt = 1.30*sin(angles(target_idx));
    text(xt, yt, '\xi^1', 'FontSize', 14, 'FontWeight', 'bold', ...
        'Color', col_target, 'HorizontalAlignment', 'center', ...
        'Interpreter', 'tex');

    % Title and annotations
    title(title_str, 'FontSize', 14, 'FontWeight', 'bold');
    xlim([-1.5 1.5]); ylim([-1.5 1.5]);

    % Count stats
    n_edges = sum(adj(:))/2;
    n_comp  = sum(comp);
    text(0, -1.40, sprintf('q_{eff}: large    edges: %d    cluster: %d/%d', ...
        n_edges, n_comp, n), 'FontSize', 10, 'HorizontalAlignment', 'center', ...
        'Color', [0.4 0.4 0.4]);
end

%% Create figure
fig = figure('Position', [50 100 1200 550], 'Color', 'w');

ax1 = subplot(1, 2, 1);
draw_panel(ax1, angles, theta_low, adj_low, comp_low, target_idx, ...
    col_target, col_connected, col_isolated, ...
    col_edge_on, col_edge_off, col_sphere, ...
    'Low temperature (retrieval)');
% Override the q_eff annotation
children = get(ax1, 'Children');
% Replace annotation
text(0, -1.40, sprintf('q_{eff} \\approx \\phi_c    edges: %d    cluster: %d/%d', ...
    sum(adj_low(:))/2, sum(comp_low), n_patterns), ...
    'FontSize', 10, 'HorizontalAlignment', 'center', ...
    'Color', [0.4 0.4 0.4], 'Parent', ax1);

ax2 = subplot(1, 2, 2);
draw_panel(ax2, angles, theta_high, adj_high, comp_high, target_idx, ...
    col_target, col_connected, col_isolated, ...
    col_edge_on, col_edge_off, col_sphere, ...
    'High temperature (paramagnetic)');
text(0, -1.40, sprintf('q_{eff} < \\phi_c    edges: %d    cluster: %d/%d', ...
    sum(adj_high(:))/2, sum(comp_high), n_patterns), ...
    'FontSize', 10, 'HorizontalAlignment', 'center', ...
    'Color', [0.4 0.4 0.4], 'Parent', ax2);

sgtitle('Basin connectivity on S^{N-1}: percolation at finite temperature', ...
    'FontSize', 16, 'FontWeight', 'bold', 'Interpreter', 'tex');

print('-dpng', '-r200', 'basin_connectivity_schematic.png');
fprintf('Saved: basin_connectivity_schematic.png\n');

File: NIPS_Resilience/code/basin_connectivity_schematic.jl
#=
Schematic: basin connectivity at low vs high temperature
Left:  low T  — basins narrow, target isolated (retrieval)
Right: high T — basins widen, giant connected cluster (paramagnetic)

Patterns are shown as disks in the 2D plane; overlapping disks are connected.

Output: basin_connectivity_schematic.png
=#

using Plots, Random, Printf

Random.seed!(7)

# ── Parameters ──
n_pat = 15
# Random 2D positions (representing pattern centers projected to 2D)
px = rand(n_pat) * 10
py = rand(n_pat) * 10

target = 1  # target pattern

# Basin radii for two temperatures
r_low  = 1.05   # low T:  small disks, few overlaps
r_high = 1.80   # high T: large disks, percolation

# ── Helpers ──
dist(i, j) = sqrt((px[i] - px[j])^2 + (py[i] - py[j])^2)

function bfs_component(adj, start)
    n = size(adj, 1)
    visited = falses(n)
    queue = [start]
    visited[start] = true
    while !isempty(queue)
        node = popfirst!(queue)
        for j in 1:n
            if adj[node, j] && !visited[j]
                visited[j] = true
                push!(queue, j)
            end
        end
    end
    return visited
end

function build_adj(r)
    adj = falses(n_pat, n_pat)
    for i in 1:n_pat, j in i+1:n_pat
        if dist(i, j) < 2r   # disks overlap
            adj[i, j] = adj[j, i] = true
        end
    end
    return adj
end

# ── Build graphs ──
adj_low  = build_adj(r_low)
adj_high = build_adj(r_high)
comp_low  = bfs_component(adj_low,  target)
comp_high = bfs_component(adj_high, target)

# ── Colors ──
col_target      = RGB(0.85, 0.22, 0.22)
col_target_fill = RGBA(0.90, 0.30, 0.30, 0.15)
col_conn        = RGB(0.25, 0.55, 0.85)
col_conn_fill   = RGBA(0.30, 0.60, 0.90, 0.12)
col_iso         = RGB(0.65, 0.65, 0.65)
col_iso_fill    = RGBA(0.70, 0.70, 0.70, 0.10)
col_edge_on     = RGBA(0.20, 0.45, 0.75, 0.6)
col_edge_off    = RGBA(0.75, 0.75, 0.75, 0.4)

# ── Draw a circle (disk outline) ──
function draw_disk!(p, cx, cy, r; linecolor=:gray, fillcolor=nothing,
                    fillalpha=0.1, lw=1.5, ls=:solid)
    θ = range(0, 2π, length=120)
    x = cx .+ r .* cos.(θ)
    y = cy .+ r .* sin.(θ)
    if fillcolor !== nothing
        plot!(p, Shape(x, y), fillcolor=fillcolor, fillalpha=fillalpha,
              linecolor=linecolor, linewidth=lw, linestyle=ls, label=false)
    else
        plot!(p, x, y, lw=lw, color=linecolor, ls=ls, label=false)
    end
end

# ── Draw one panel ──
function draw_panel!(p, r, adj, comp;
                     title_str="", subtitle_str="")
    # 1) Basin disks (draw first so edges and dots are on top)
    for i in 1:n_pat
        if i == target
            fc, lc = col_target_fill, col_target
        elseif comp[i]
            fc, lc = col_conn_fill, col_conn
        else
            fc, lc = col_iso_fill, col_iso
        end
        lw = i == target ? 2.5 : 1.5
        ls = comp[i] || i == target ? :solid : :dash
        draw_disk!(p, px[i], py[i], r; linecolor=lc, fillcolor=fc,
                   fillalpha=0.15, lw=lw, ls=ls)
    end

    # 2) Connection edges
    for i in 1:n_pat, j in i+1:n_pat
        if adj[i, j]
            ec = (comp[i] && comp[j]) ? col_edge_on : col_edge_off
            lw = (comp[i] && comp[j]) ? 2.0 : 1.0
            plot!(p, [px[i], px[j]], [py[i], py[j]],
                  lw=lw, color=ec, label=false)
        end
    end

    # 3) Pattern center dots
    for i in 1:n_pat
        mc = i == target ? col_target :
             comp[i]     ? col_conn : col_iso
        ms = i == target ? 8 : 6
        msw = i == target ? 2.0 : 1.5
        scatter!(p, [px[i]], [py[i]], ms=ms, color=mc,
                 markerstrokecolor=:white, markerstrokewidth=msw,
                 label=false)
    end

    # 4) Label target
    annotate!(p, px[target] + r * 0.6, py[target] + r * 0.6,
              text("ξ¹", 13, col_target))

    # 5) Stats
    n_edges = count(adj) ÷ 2
    n_comp  = count(comp)
    annotate!(p, 5.0, -1.0,
        text(subtitle_str * @sprintf("  edges: %d   cluster: %d/%d",
             n_edges, n_comp, n_pat), 9, :gray40))

    title!(p, title_str)
end

# ── Compose ──
margin = 2.0
p1 = plot(aspect_ratio=:equal,
          xlims=(minimum(px)-r_high-margin, maximum(px)+r_high+margin),
          ylims=(minimum(py)-r_high-margin, maximum(py)+r_high+margin),
          axis=false, grid=false, framestyle=:none, legend=false,
          titlefontsize=13)
draw_panel!(p1, r_low, adj_low, comp_low;
            title_str="Low T  (retrieval)",
            subtitle_str="q_eff ≈ φ_c  ")

p2 = plot(aspect_ratio=:equal,
          xlims=(minimum(px)-r_high-margin, maximum(px)+r_high+margin),
          ylims=(minimum(py)-r_high-margin, maximum(py)+r_high+margin),
          axis=false, grid=false, framestyle=:none, legend=false,
          titlefontsize=13)
draw_panel!(p2, r_high, adj_high, comp_high;
            title_str="High T  (paramagnetic)",
            subtitle_str="q_eff < φ_c  ")

fig = plot(p1, p2, layout=(1, 2), size=(1400, 650), dpi=150,
           plot_title="Basin percolation: temperature drives connectivity",
           margin=5Plots.mm)

savefig(fig, "basin_connectivity_schematic.png")
println("✓ Saved: basin_connectivity_schematic.png")
